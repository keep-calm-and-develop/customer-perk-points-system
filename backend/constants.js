const BASE_THRESHOLD = 50;
const BASE_MULTIPLIER = 1;

const TOP_THRESHOLD = 100;
const TOP_MULTIPLIER = 2;

const API_RESPONSE_CODES = {
    SUCCESS: 200,
    SERVER_ERROR: 500,
    NOT_FOUND: 404,
    INVALID_REQUEST: 400,
};

const NANO_ID_SIZE = 4;

const FIXED_DECIMAL_POSITIONS = 2;

const MONTHS = [
    0,
    1,
    2,
];

export {
    BASE_THRESHOLD,
    BASE_MULTIPLIER,
    TOP_THRESHOLD,
    TOP_MULTIPLIER,
    API_RESPONSE_CODES,
    NANO_ID_SIZE,
    FIXED_DECIMAL_POSITIONS,
    MONTHS,
};
