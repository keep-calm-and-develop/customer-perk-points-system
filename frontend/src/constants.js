
const API_SERVER_URL = `http://localhost:3080/api`;

const API_SERVICES = {
    GET_CUSTOMERS: `${API_SERVER_URL}/customers`,
    GET_REWARD_POINTS: `${API_SERVER_URL}/customer/reward-points`,
    ADD_CUSTOMER: `${API_SERVER_URL}/customer`,
    ADD_TRANSACTION: `${API_SERVER_URL}/customer/transaction/`,
};

const MONTHS = ['Jan', 'Feb', 'March', 'April', 'May', 'June', 'July', 'Aug', 'Sept', 'Oct', 'Nov', 'Dec'];

const MOCK_CUSTOMERS = [
    {
        "name": "user1",
        "customerID": "1",
        "totalPoints": 200,
        "monthlyPoints": [150, 30, 20]
    },
    {
        "name": "user2",
        "customerID": "2",
        "totalPoints": 100,
        "monthlyPoints": [50, 30, 20]
    }
];

export {
    API_SERVER_URL,
    API_SERVICES,
    MONTHS,
    MOCK_CUSTOMERS,
};
