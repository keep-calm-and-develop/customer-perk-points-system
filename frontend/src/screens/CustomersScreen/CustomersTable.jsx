import PropTypes from "prop-types";
import { React, useMemo } from "react";
import CustomerTableRow from "./CustomerTableRow";
import getLastThreeMonths from "./getLastThreeMonths";

export default function CustomersTable({ customers }) {
    const lastThreeMonths = useMemo(() => getLastThreeMonths(), []);
    return (
        customers.length === 0
            ? <h3 className="no-customers-message">0 Customers</h3>
            : <table className="customers-table">
                <thead>
                    <tr>
                        <th>Name</th>
                        <th>Total Points</th>
                        {lastThreeMonths.map((month) => <th key={month}>{month}</th>)}
                        <th>Transaction Action</th>
                    </tr>
                </thead>
                <tbody>
                    {customers.map(({ name, customerID, totalPoints, monthlyPoints }) => (
                        <CustomerTableRow
                            key={customerID}
                            name={name}
                            customerID={customerID}
                            totalPoints={totalPoints}
                            monthlyPoints={monthlyPoints}
                        />
                    ))}
                </tbody>
            </table>
    );
}

CustomersTable.propTypes = {
    customers: PropTypes.arrayOf(PropTypes.shape({
        name: PropTypes.string,
        customerID: PropTypes.string,
        totalPoints: PropTypes.number,
        monthlyPoints: PropTypes.arrayOf(PropTypes.number),
    })).isRequired,
}
