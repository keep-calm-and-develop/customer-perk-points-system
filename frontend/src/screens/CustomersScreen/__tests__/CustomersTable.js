import { cleanup, render, screen, waitFor } from "@testing-library/react";
import React from "react";
import { act } from "react-dom/test-utils";
import { BrowserRouter } from "react-router-dom";
import { MOCK_CUSTOMERS } from "../../../constants";
import CustomersTable from "../CustomersTable";

afterEach(cleanup);

describe('<CustomersTable />', () => {
    it('Should match snapshot', () => {
        const customersTable = render(
            <BrowserRouter>
                <CustomersTable customers={MOCK_CUSTOMERS} />
            </BrowserRouter>
        );
        expect(customersTable).toMatchSnapshot();
    });

    it('Should render 2 rows', async () => {
        await act(async () => render(
            <BrowserRouter>
                <CustomersTable customers={MOCK_CUSTOMERS} />
            </BrowserRouter>
        ));
        const customerRows = await waitFor(() => screen.findAllByTestId("customer"));

        expect(customerRows).toHaveLength(2);
    });

    it('Should render text 0 customers', async () => {
        await act(async () => render(
            <BrowserRouter>
                <CustomersTable customers={[]} />
            </BrowserRouter>
        ));
        const noCustomersText = screen.getByText(/0 customers/i);

        expect(noCustomersText).toBeInTheDocument();
    });
});

