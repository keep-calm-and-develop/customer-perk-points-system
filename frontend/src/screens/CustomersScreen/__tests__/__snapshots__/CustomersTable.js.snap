// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomersTable /> Should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        class="customers-table"
        data-testid="customers-table"
      >
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Total Points
            </th>
            <th>
              Jan
            </th>
            <th>
              Dec
            </th>
            <th>
              Nov
            </th>
            <th>
              Transaction Action
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="customers-list__row"
            data-testid="customer"
          >
            <td
              class="customers-list__cell-name"
              data-testid="customer-name"
            >
              user1
            </td>
            <td
              class="customers-list__cell-points"
              data-testid="customer-total-points"
            >
              200
            </td>
            <td
              class="customers-list__cell-points"
              data-testid="customer-month-0-points"
            >
              150
            </td>
            <td
              class="customers-list__cell-points"
              data-testid="customer-month-1-points"
            >
              30
            </td>
            <td
              class="customers-list__cell-points"
              data-testid="customer-month-2-points"
            >
              20
            </td>
            <td
              data-testid="customer-add-transaction"
            >
              <a
                class="customers-list__cell-add-transaction"
                href="/1/new-transaction?name=user1"
              >
                Add Transaction
              </a>
            </td>
          </tr>
          <tr
            class="customers-list__row"
            data-testid="customer"
          >
            <td
              class="customers-list__cell-name"
              data-testid="customer-name"
            >
              user2
            </td>
            <td
              class="customers-list__cell-points"
              data-testid="customer-total-points"
            >
              100
            </td>
            <td
              class="customers-list__cell-points"
              data-testid="customer-month-0-points"
            >
              50
            </td>
            <td
              class="customers-list__cell-points"
              data-testid="customer-month-1-points"
            >
              30
            </td>
            <td
              class="customers-list__cell-points"
              data-testid="customer-month-2-points"
            >
              20
            </td>
            <td
              data-testid="customer-add-transaction"
            >
              <a
                class="customers-list__cell-add-transaction"
                href="/2/new-transaction?name=user2"
              >
                Add Transaction
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      class="customers-table"
      data-testid="customers-table"
    >
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Total Points
          </th>
          <th>
            Jan
          </th>
          <th>
            Dec
          </th>
          <th>
            Nov
          </th>
          <th>
            Transaction Action
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="customers-list__row"
          data-testid="customer"
        >
          <td
            class="customers-list__cell-name"
            data-testid="customer-name"
          >
            user1
          </td>
          <td
            class="customers-list__cell-points"
            data-testid="customer-total-points"
          >
            200
          </td>
          <td
            class="customers-list__cell-points"
            data-testid="customer-month-0-points"
          >
            150
          </td>
          <td
            class="customers-list__cell-points"
            data-testid="customer-month-1-points"
          >
            30
          </td>
          <td
            class="customers-list__cell-points"
            data-testid="customer-month-2-points"
          >
            20
          </td>
          <td
            data-testid="customer-add-transaction"
          >
            <a
              class="customers-list__cell-add-transaction"
              href="/1/new-transaction?name=user1"
            >
              Add Transaction
            </a>
          </td>
        </tr>
        <tr
          class="customers-list__row"
          data-testid="customer"
        >
          <td
            class="customers-list__cell-name"
            data-testid="customer-name"
          >
            user2
          </td>
          <td
            class="customers-list__cell-points"
            data-testid="customer-total-points"
          >
            100
          </td>
          <td
            class="customers-list__cell-points"
            data-testid="customer-month-0-points"
          >
            50
          </td>
          <td
            class="customers-list__cell-points"
            data-testid="customer-month-1-points"
          >
            30
          </td>
          <td
            class="customers-list__cell-points"
            data-testid="customer-month-2-points"
          >
            20
          </td>
          <td
            data-testid="customer-add-transaction"
          >
            <a
              class="customers-list__cell-add-transaction"
              href="/2/new-transaction?name=user2"
            >
              Add Transaction
            </a>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
